Mathlib/CategoryTheory/Limits/HasLimits.lean

```diff
@@ -0,0 +1,1355 @@
+
+/-- `HasLimit F` represents the mere existence of a limit for `F`. -/
+class HasLimit (F : J ⥤ C) : Prop where mk' ::
+  /-- There is some limit cone for `F` -/
+  exists_limit : Nonempty (LimitCone F)
+#align category_theory.limits.has_limit CategoryTheory.Limits.HasLimit
+
+theorem HasLimit.mk {F : J ⥤ C} (d : LimitCone F) : HasLimit F :=
+  ⟨Nonempty.intro d⟩
+#align category_theory.limits.has_limit.mk CategoryTheory.Limits.HasLimit.mk
+
+/-- Use the axiom of choice to extract explicit `LimitCone F` from `HasLimit F`. -/
+def getLimitCone (F : J ⥤ C) [HasLimit F] : LimitCone F :=
+  Classical.choice <| HasLimit.exists_limit
+#align category_theory.limits.get_limit_cone CategoryTheory.Limits.getLimitCone
+
+variable (J C)
+
+/-- `C` has limits of shape `J` if there exists a limit for every functor `F : J ⥤ C`. -/
+class HasLimitsOfShape : Prop where
+  /-- All functors `F : J ⥤  C` from `J` have limits -/
+  has_limit : ∀ F : J ⥤ C, HasLimit F := by infer_instance
+#align category_theory.limits.has_limits_of_shape CategoryTheory.Limits.HasLimitsOfShape
+
+/-- `C` has all limits of size `v₁ u₁` (`HasLimitsOfSize.{v₁ u₁} C`)
+if it has limits of every shape `J : Type u₁` with `[Category.{v₁} J]`.
+-/
+class HasLimitsOfSize (C : Type u) [Category.{v} C] : Prop where
+  /-- All functors `F : J ⥤ C` from all small `J` have limits -/
+  has_limits_of_shape : ∀ (J : Type u₁) [Category.{v₁} J], HasLimitsOfShape J C := by
+    infer_instance
+#align category_theory.limits.has_limits_of_size CategoryTheory.Limits.HasLimitsOfSize
+
+/-- `C` has all (small) limits if it has limits of every shape that is as big as its hom-sets. -/
+abbrev HasLimits (C : Type u) [Category.{v} C] : Prop :=
+  HasLimitsOfSize.{v, v} C
+#align category_theory.limits.has_limits CategoryTheory.Limits.HasLimits
+
+theorem HasLimits.has_limits_of_shape {C : Type u} [Category.{v} C] [HasLimits C] (J : Type v)
+    [Category.{v} J] : HasLimitsOfShape J C :=
+  HasLimitsOfSize.has_limits_of_shape J
+#align category_theory.limits.has_limits.has_limits_of_shape CategoryTheory.Limits.HasLimits.has_limits_of_shape
+
+variable {J C}
+
+-- see Note [lower instance priority]
+instance (priority := 100) hasLimitOfHasLimitsOfShape {J : Type u₁} [Category.{v₁} J]
+    [HasLimitsOfShape J C] (F : J ⥤ C) : HasLimit F :=
+  HasLimitsOfShape.has_limit F
+#align category_theory.limits.has_limit_of_has_limits_of_shape CategoryTheory.Limits.hasLimitOfHasLimitsOfShape
+
+-- see Note [lower instance priority]
+instance (priority := 100) hasLimitsOfShapeOfHasLimits {J : Type u₁} [Category.{v₁} J]
+    [HasLimitsOfSize.{v₁, u₁} C] : HasLimitsOfShape J C :=
+  HasLimitsOfSize.has_limits_of_shape J
+#align category_theory.limits.has_limits_of_shape_of_has_limits CategoryTheory.Limits.hasLimitsOfShapeOfHasLimits
+
+-- Interface to the `HasLimit` class.
+/-- An arbitrary choice of limit cone for a functor. -/
+def Limit.cone (F : J ⥤ C) [HasLimit F] : Cone F :=
```

Comment:
Is the naming correct here? `Limit.cone` vs `limit.lift` and friends.

Reply:
No I don't think so

Reply:
Made the change